test/testers.jl

```diff
@@ -0,0 +1,540 @@
+    rtol::Real=1e-9,
+    atol::Real=1e-9,
+    testset_name=nothing,
+)
+    call_with_copy(f, xs...) = deepcopy(f)(deepcopy(xs)...; deepcopy(fkwargs)...)
+    call_with_kwargs(f, xs...) = f(xs...; fkwargs...)
+    if testset_name === nothing
+        testset_name = "test_forward: $(f isa Const ? f.val : f) with return activity $ret_activity on $(args)"
+    end
+    @testset "$testset_name" begin
+        # format arguments for autodiff and FiniteDifferences
+        activities = map(auto_forward_activity, (f, args...))
+        primals = map(x -> x.val, activities)
+        # call primal, avoid mutating original arguments
+        y = call_with_copy(primals...)
+        # TODO: handle batch activities
+        # call finitedifferences, avoid mutating original arguments
+        dy_fdm = _make_jvp_call(fdm, call_with_copy, ret_activity, y, activities)
+        # call autodiff, allow mutating original arguments
+        y_and_dy_ad = autodiff(Forward, call_with_kwargs, ret_activity, activities...)
+        if ret_activity <: Union{Duplicated,BatchDuplicated}
+            y_ad, dy_ad = y_and_dy_ad
+            # check primal agrees with primal function
+            test_approx(y_ad, y; atol, rtol)
+        elseif ret_activity <: Union{DuplicatedNoNeed,BatchDuplicatedNoNeed}
+            # check primal is not returned
+            @test length(y_and_dy_ad) == 1
+            dy_ad = y_and_dy_ad[1]
+        elseif ret_activity <: Const
+            # check Const activity returns an empty tuple
+            @test isempty(y_and_dy_ad)
+            dy_ad = ()
+        else
+            throw(ArgumentError("Unsupported return activity type: $ret_activity"))
+        end
+        if y isa Tuple
+            # check Enzyme and FiniteDifferences return the same number of derivatives
+            @test length(dy_ad) == length(dy_fdm)
+            # check all returned derivatives against FiniteDifferences
+            for (dy_ad_i, dy_fdm_i) in zip(dy_ad, dy_fdm)
+                test_approx(dy_ad_i, dy_fdm_i; atol, rtol)
+            end
+        else
+            test_approx(dy_ad, dy_fdm; atol, rtol)
+        end
+    end
+end
+
+"""
+    all_or_no_batch(activities...) -> Bool
+
+Returns `true` if `activities` are compatible in terms of batched activities.
+
+When a test set loops over many activities, some of which may be `BatchedDuplicated` or
+`BatchedDuplicatedNoNeed`, this is useful for skipping those combinations that are
+incompatible and will raise errors.
+"""
+function all_or_no_batch(activities...)
+    no_batch = !any(activities) do T
+        T <: Union{BatchDuplicated,BatchDuplicatedNoNeed}
+    end
+    all_batch_or_const = all(activities) do T
+        T <: Union{BatchDuplicated,BatchDuplicatedNoNeed,Const}
+    end
+    return all_batch_or_const || no_batch
+end
+
+# meta-testing functions adapted from https://github.com/JuliaDiff/ChainRulesTestUtils.jl/blob/v1.10.1/test/meta_testing_tools.jl
+# Copyright (c) 2020 JuliaDiff
```

Comment:
These are just here temporarily and will be removed once https://github.com/sethaxen/MetaTesting.jl is registered.